### PR TITLE
compaction: introduce multilevel compaction heuristic interface

### PR DESCRIPTION
**Pull request by msbutler**
This patch introduces the multiLevelHueristic interface, whose implementations
can evaluate whether to add an additional level to a conventional 2 level
compaction.

If the option.experimental.multiLevelHeuristic knob is set, the compaction
picker will create a deep copy of the original compaction, call setupInputs on
the newly allocated object, adding a new level to the new compaction
candidate. Then, the heuristic decides to compact with the original picked
compaction or the new multilevel candidate.

Informs #2053

### CODE REVIEW

**Comment by msbutler** [comment]
Hoping to get this in now to prevent future merge conflicts and to speed up development while experimenting with various multilevel compaction heuristics.

**Comment by msbutler** [comment]
huh, github is allowing me to merge this PR without a review. Is this expected?

**Review by itsbilal — COMMENTED** [review]
Looks generally good but feels a little odd to get this in without a single test using it? Maybe you could add a quick test to `compaction_setup_inputs` that shows this in action.

> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/2129#-:-NI8rzSL4Ff9mCmB-ub7:br7w8vv)* status: all files reviewed, 2 unresolved discussions (waiting on @msbutler)

___
*[`compaction_picker_test.go` line 1009 at r1](https://reviewable.io/reviews/cockroachdb/pebble/2129#-NI8qgKH79n9h0FpRJkU:-NI8qgKI5ZVw8G7VHFZj:b-q5u6ht) ([raw file](https://github.com/cockroachdb/pebble/blob/9e337c422db986fbbf262fbc20a9b6f45bc3a12c/compaction_picker_test.go#L1009)):*
> ```Go
> 				switch data[:2] {
> 				case "L0", "L1", "L2", "L3", "L4", "L5", "L6":
> 					levelArgs := strings.Split(data, " ")
> ```

Nit: we usually use `strings.Fields` for this, as it has the benefit of handling different types of whitespace characters the same way.
___
*[`compaction_picker_test.go` line 1075 at r1](https://reviewable.io/reviews/cockroachdb/pebble/2129#-NI8ts5KDrxbhMrCsinp:-NI8ts5KDrxbhMrCsinq:bmze3y5) ([raw file](https://github.com/cockroachdb/pebble/blob/9e337c422db986fbbf262fbc20a9b6f45bc3a12c/compaction_picker_test.go#L1075)):*
> ```Go
> 			// Scaffolding for running a multi level compaction test
> 			var initMultiLevel bool
> 			if initMultiLevel {
> ```

This will always be false, right?

**Comment by msbutler** [comment]
@itsbilal RFAL. Reviewable does not seem to be working for me rn, mind if we review over github?

I added a bit more to this PR and address your initial comments. Happy to walk through this over a call!

**Review by itsbilal — COMMENTED** [review]
Sorry for the late review but i'm really glad to see this take shape! No major comments, just one question about where you expect the implementation of heuristics to live, and the effects that has on some other things in here.

> Reviewed 7 of 7 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/2129#-:-NLDbIMqBrOrdQT-Zx4m:bes1jiw)* status: all files reviewed, 5 unresolved discussions (waiting on @msbutler)

___
*[`compaction_picker.go` line 276 at r2](https://reviewable.io/reviews/cockroachdb/pebble/2129#-NLD_ALU13pn6it2nsp8:-NLD_ALU13pn6it2nsp9:b-7nw6c0) ([raw file](https://github.com/cockroachdb/pebble/blob/24e5851f2ef82b6b979db6309552e6d4f2e9e60f/compaction_picker.go#L276)):*
> ```Go
> 		score:                  pc.score,
> 		kind:                   pc.kind,
> 		startLevel:             nil,
> ```

nit: we usually eliminate zero-value fields from literals, so these `nil` fields can just be removed.
___
*[`compaction_picker.go` line 1540 at r2](https://reviewable.io/reviews/cockroachdb/pebble/2129#-NLD_Zh58GXpIY66CoTW:-NLD_Zh58GXpIY66CoTX:b-5adw95) ([raw file](https://github.com/cockroachdb/pebble/blob/24e5851f2ef82b6b979db6309552e6d4f2e9e60f/compaction_picker.go#L1540)):*
> ```Go
> 
> // maybeAddLevel maybe adds a level to the picked compaction.
> func maybeAddLevel(pc *pickedCompaction, opts *Options, diskAvailBytes uint64) *pickedCompaction {
> ```

We'd probably wanna make `pc` a receiver here:

```
func (pc *pickedCompaction) maybeAddLevel(opts *Options, diskAvailBytes uint64) *pickedCompaction {
```
___
*[`compaction_picker.go` line 1563 at r2](https://reviewable.io/reviews/cockroachdb/pebble/2129#-NLDaMocEij15fv2JDag:-NLDaMocEij15fv2JDah:b1vfsu2) ([raw file](https://github.com/cockroachdb/pebble/blob/24e5851f2ef82b6b979db6309552e6d4f2e9e60f/compaction_picker.go#L1563)):*
> ```Go
> type multiLevelHueristic interface {
> 	// addLevel returns the preferred compaction.
> 	addLevel(pcAddedLevel *pickedCompaction, pcOrig *pickedCompaction) *pickedCompaction
> ```

I wonder what the motivation behind making this an interface was. Is it mostly to isolate the mechanics in this file away from the math of it all? And do you expect the implementations to live in this package or outside of it?

I like the idea to keep the mathematical heuristic separate from the mechanics, however it might be too painful to do a heuristic implementation in, say, Cockroach since a lot of `pickedCompaction` fields/methods if not all of them are unexported. It might make more sense to have a `MultiLevelHeuristicDefault` or so that we export, that we could set the Option to it if we want to enable multi-level compactions, and the heuristic itself lives in this package and can see the inside of `compactionPicker no problem. Is that what you had in mind?

If the heuristics are going to live inside this package, then it might actually be a good idea to have the heuristic just take `pcOrig`. The heuristic can be responsible for calling Clone and `pc.setupMultiLevelCandidate` if it decides that it's worthwhile, however it doesn't need to waste that work until it has made that determination. This would also clean up the interface. Thoughts?

**Review by msbutler — COMMENTED** [review]
Apologies for the slow reply! Day job was crazy this week. Free all day today to go back and forth on this!

I may add a few unit tests to these clone methods as well while you mull over my comments below.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/2129#-:-NLfc2l90mr0wpt6yH7N:bdmich0)* status: all files reviewed, 3 unresolved discussions (waiting on @itsbilal)

___
*[`compaction_picker.go` line 276 at r2](https://reviewable.io/reviews/cockroachdb/pebble/2129#-NLD_ALU13pn6it2nsp8:-NLfMlE77mHgd49lty4G:bb6un1j) ([raw file](https://github.com/cockroachdb/pebble/blob/24e5851f2ef82b6b979db6309552e6d4f2e9e60f/compaction_picker.go#L276)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

nit: we usually eliminate zero-value fields from literals, so these `nil` fields can just be removed.
</blockquote></details>

Done
___
*[`compaction_picker.go` line 1540 at r2](https://reviewable.io/reviews/cockroachdb/pebble/2129#-NLD_Zh58GXpIY66CoTW:-NLfc0gL5ELhexfDY3x3:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/24e5851f2ef82b6b979db6309552e6d4f2e9e60f/compaction_picker.go#L1540)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

We'd probably wanna make `pc` a receiver here:

```
func (pc *pickedCompaction) maybeAddLevel(opts *Options, diskAvailBytes uint64) *pickedCompaction {
```
</blockquote></details>

Done.
___
*[`compaction_picker.go` line 1563 at r2](https://reviewable.io/reviews/cockroachdb/pebble/2129#-NLDaMocEij15fv2JDag:-NLfPTJY7b4qnN_PBh2-:b7vvm75) ([raw file](https://github.com/cockroachdb/pebble/blob/24e5851f2ef82b6b979db6309552e6d4f2e9e60f/compaction_picker.go#L1563)):*
> Why an interface?
I mades this an interface so it would be easy to create other multiLevel heuristics. Given that there's only one for now, we don't need to make this an interface. 

> What to expose to the cockroach binary?
Yeah,  agree that the implementation of the interface should not live in cockroach. The engine user should merely pick which heuristic they want (e.g. none or writeAmp), and not implement it. 

> Should the multiLevelHeuristic instead call `Clone` and `SetupMultiLevelCandidate`?
hmmm, I'm not sure. I view the `Clone` and `SetupMultiLevelCandidate` as mechanical tasks that seem separate from the heuristic evaluation. But you are right that it does clean up the impl. of `maybeAddLevel`.  I've implemented the change in this the added commit, and I think I like it heh.
___
*[`compaction_picker_test.go` line 1009 at r1](https://reviewable.io/reviews/cockroachdb/pebble/2129#-NI8qgKH79n9h0FpRJkU:-NImrF-J1K879TAPfenA:b-896fix) ([raw file](https://github.com/cockroachdb/pebble/blob/9e337c422db986fbbf262fbc20a9b6f45bc3a12c/compaction_picker_test.go#L1009)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

Nit: we usually use `strings.Fields` for this, as it has the benefit of handling different types of whitespace characters the same way.
</blockquote></details>

Done.
___
*[`compaction_picker_test.go` line 1075 at r1](https://reviewable.io/reviews/cockroachdb/pebble/2129#-NI8ts5KDrxbhMrCsinp:-NImrFpm9ZDthm61mtlm:bcvq78a) ([raw file](https://github.com/cockroachdb/pebble/blob/9e337c422db986fbbf262fbc20a9b6f45bc3a12c/compaction_picker_test.go#L1075)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

This will always be false, right?
</blockquote></details>

No Longer relevant

**Comment by msbutler** [comment]
@itsbilal just added another commit which adds a function to check compaction cloning correctness. Do y'all have some sort of env var or setting used to conduct extra checks during large scale tests? It would be great to randomly clone and check picked compactions out in the wild, using this `checkClone` func I just added.

**Review by itsbilal — APPROVED** [review]
:lgtm: thanks so much for working on this! Only remaining comments are nits, you can address them and merge without waiting for a re-review.

And to answer your question, we have the `-tags 'invariants'` flag you can pass to `go test`, which enables additional invariant checking in the code for all sorts of correctness cases. that + running the test suite (esp. the metamorphic tests) should be pretty sufficient!

> Reviewed 1 of 5 files at r3, 2 of 4 files at r4, 4 of 4 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/2129#-:-NMAjYYjB7Ev0jd7nja6:b-o9e7h4)* status: all files reviewed, 5 unresolved discussions (waiting on @msbutler)

___
*[`compaction_picker.go` line 1563 at r2](https://reviewable.io/reviews/cockroachdb/pebble/2129#-NLDaMocEij15fv2JDag:-NMAlEAB2nsDY2IgTJc1:b-nlu8r2) ([raw file](https://github.com/cockroachdb/pebble/blob/24e5851f2ef82b6b979db6309552e6d4f2e9e60f/compaction_picker.go#L1563)):*
<details><summary><i>Previously, msbutler (Michael Butler) wrote…</i></summary><blockquote>

> Why an interface?
I mades this an interface so it would be easy to create other multiLevel heuristics. Given that there's only one for now, we don't need to make this an interface. 

> What to expose to the cockroach binary?
Yeah,  agree that the implementation of the interface should not live in cockroach. The engine user should merely pick which heuristic they want (e.g. none or writeAmp), and not implement it. 

> Should the multiLevelHeuristic instead call `Clone` and `SetupMultiLevelCandidate`?
hmmm, I'm not sure. I view the `Clone` and `SetupMultiLevelCandidate` as mechanical tasks that seem separate from the heuristic evaluation. But you are right that it does clean up the impl. of `maybeAddLevel`.  I've implemented the change in this the added commit, and I think I like it heh.
</blockquote></details>

Looks great now, thanks!
___
*[`compaction_picker.go` line 273 at r5](https://reviewable.io/reviews/cockroachdb/pebble/2129#-NMAkmkp2QW1OJQh1GTk:-NMAkmkp2QW1OJQh1GTl:b-utdpg1) ([raw file](https://github.com/cockroachdb/pebble/blob/fd4a9e63b68510bb5e6e6b20b352493dd5b22a5e/compaction_picker.go#L273)):*
> ```Go
> func (pc *pickedCompaction) String() string {
> 	var builder strings.Builder
> 	builder.WriteString(fmt.Sprintf(`Score=%f\n`, pc.score))
> ```

nit: maybe we should use a delimeter other than \n here as it's unusual to return a multiline string for a `String()` method. Maybe commas `,` ?
___
*[`compaction_picker.go` line 1577 at r5](https://reviewable.io/reviews/cockroachdb/pebble/2129#-NMAlHqw9LZo4mXakAEe:-NMAlHqw9LZo4mXakAEf:bsyvori) ([raw file](https://github.com/cockroachdb/pebble/blob/fd4a9e63b68510bb5e6e6b20b352493dd5b22a5e/compaction_picker.go#L1577)):*
> ```Go
> type MultiLevelHeuristic interface {
> 	// evaluate returns the preferred compaction.
> 	evaluate(pc *pickedCompaction, opts *Options, diskAvailBytes uint64) *pickedCompaction
> ```

nit: maybe `pickCompaction` or `pick` is a better function name than `evaluate`, that way it's also more in line with the `compactionPicker` interface above.
___
*[`compaction_picker.go` line 1582 at r5](https://reviewable.io/reviews/cockroachdb/pebble/2129#-NMAkxWe-96lD__rfSw8:-NMAkxWe-96lD__rfSw9:bleop29) ([raw file](https://github.com/cockroachdb/pebble/blob/fd4a9e63b68510bb5e6e6b20b352493dd5b22a5e/compaction_picker.go#L1582)):*
> ```Go
> // NoMultiLevel will never add an additional level to the compaction.
> type NoMultiLevel struct {
> }
> ```

nit: `type NoMultiLevel struct{}` instead of multiline.
___
*[`internal/manifest/l0_sublevels.go` line 1164 at r5](https://reviewable.io/reviews/cockroachdb/pebble/2129#-NMAlWuY62MJJdTiH3oq:-NMAlWuY62MJJdTiH3or:b2girma) ([raw file](https://github.com/cockroachdb/pebble/blob/fd4a9e63b68510bb5e6e6b20b352493dd5b22a5e/internal/manifest/l0_sublevels.go#L1164)):*
> ```Go
> func (l *L0CompactionFiles) String() string {
> 	if len(l.Files) > 0 {
> 		return fmt.Sprintf("First File Address: %p", &l.Files[0])
> ```

Is this useful? Might be better to print the filenum of the first file, i.e. `l.Files[0].FileNum.String()`.

**Review by msbutler — COMMENTED** [review]
Sweet! I will plan to experiment with those test tags once this PR merges.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/2129#-:-NMEf8bP4A0_FkzPlNUW:b-m7a9nu)* status: all files reviewed, 2 unresolved discussions (waiting on @itsbilal)

___
*[`compaction_picker.go` line 273 at r5](https://reviewable.io/reviews/cockroachdb/pebble/2129#-NMAkmkp2QW1OJQh1GTk:-NMEf3CO2Ow9QoNELwiF:bb6un1j) ([raw file](https://github.com/cockroachdb/pebble/blob/fd4a9e63b68510bb5e6e6b20b352493dd5b22a5e/compaction_picker.go#L273)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

nit: maybe we should use a delimeter other than \n here as it's unusual to return a multiline string for a `String()` method. Maybe commas `,` ?
</blockquote></details>

Done
___
*[`internal/manifest/l0_sublevels.go` line 1164 at r5](https://reviewable.io/reviews/cockroachdb/pebble/2129#-NMAlWuY62MJJdTiH3oq:-NMEd5yb4sQTNAdgFvzh:b85ewf2) ([raw file](https://github.com/cockroachdb/pebble/blob/fd4a9e63b68510bb5e6e6b20b352493dd5b22a5e/internal/manifest/l0_sublevels.go#L1164)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

Is this useful? Might be better to print the filenum of the first file, i.e. `l.Files[0].FileNum.String()`.
</blockquote></details>

Aha, I chose to print the addr bc I want to assert that we're doing a shallow copy of the contents of L0CompactionFiles. Will keep this as is for now.

**Pull request merged**
